feature: add terminal_state metrics

googs1025 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
- I found that only counter seems to be insufficient info, we should add Gauge metrics for jobset. So that we can query for a single jobset terminal status
#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
add jobset_terminal_state  metrics
```

googs1025 commented:
- [ ] todo: update docs

googs1025 commented on pkg/controllers/jobset_controller.go:
```diff
@@ -271,6 +272,25 @@ func (r *JobSetReconciler) updateJobSetStatus(ctx context.Context, js *jobset.Jo
+// RecordJobSetTerminalState records the terminal state of a JobSet.
+// This function checks the terminal state of the given JobSet and records metrics for it.
+// If the terminal state is empty, no metrics are recorded.
+// Otherwise, it records whether the terminal state matches "Completed" or "Failed".
```
some Inspired came from:

https://github.com/kubernetes/kube-state-metrics/blob/3aee5ff2bbefc8b882d744edb69064c10ac9e50e/internal/store/pod.go#L1336

kannon92 commented:
Code looks good to me.

I'll wait until docs are there.

kannon92 commented:
Testing copilot feature for fun..

This pull request introduces functionality to record and track the terminal state of JobSets using Prometheus metrics. The changes include adding a new metric, implementing the recording logic, and adding tests to ensure the functionality works as expected.

### Metrics and Recording:

* [`pkg/metrics/metrics.go`](diffhunk://#diff-4bc91a74bc47b8467cdcd25a7e3fe1651dbe44907f478377eca31b982f444f23R42-R49): Added a new Prometheus gauge metric `JobSetTerminalState` to track the terminal state of JobSets. Implemented the `RecordJobSetTerminalState` function to set the value of this metric based on the JobSet's terminal state. [[1]](diffhunk://#diff-4bc91a74bc47b8467cdcd25a7e3fe1651dbe44907f478377eca31b982f444f23R42-R49) [[2]](diffhunk://#diff-4bc91a74bc47b8467cdcd25a7e3fe1651dbe44907f478377eca31b982f444f23R64-R79)

### Controller Update:

* [`pkg/controllers/jobset_controller.go`](diffhunk://#diff-68677f8530f00fbbb399ac36268f928ee584948dd09091644a8fad81a8658153R259): Updated the `updateJobSetStatus` function to call `RecordJobSetTerminalState` when updating the JobSet status. Added the `RecordJobSetTerminalState` function to check and record the terminal state of a JobSet. [[1]](diffhunk://#diff-68677f8530f00fbbb399ac36268f928ee584948dd09091644a8fad81a8658153R259) [[2]](diffhunk://#diff-68677f8530f00fbbb399ac36268f928ee584948dd09091644a8fad81a8658153R275-R293)

### Tests:

* [`pkg/metrics/metrics_test.go`](diffhunk://#diff-c3ade6c0a09b6258d971210f5a41c2ae5beb77873cee44a9ccb78055caf271ddR65-R84): Added a new test `TestRecordJobSetTerminalState` to verify that the `RecordJobSetTerminalState` function correctly sets the gauge metric for different terminal states.

googs1025 commented:
> Testing copilot feature for fun..
> 
> This pull request introduces functionality to record and track the terminal state of JobSets using Prometheus metrics. The changes include adding a new metric, implementing the recording logic, and adding tests to ensure the functionality works as expected.
> 
> ### Metrics and Recording:
> * [`pkg/metrics/metrics.go`](https://github.com/kubernetes-sigs/jobset/pull/872/files#diff-4bc91a74bc47b8467cdcd25a7e3fe1651dbe44907f478377eca31b982f444f23R42-R49): Added a new Prometheus gauge metric `JobSetTerminalState` to track the terminal state of JobSets. Implemented the `RecordJobSetTerminalState` function to set the value of this metric based on the JobSet's terminal state. [[1]](https://github.com/kubernetes-sigs/jobset/pull/872/files#diff-4bc91a74bc47b8467cdcd25a7e3fe1651dbe44907f478377eca31b982f444f23R42-R49) [[2]](https://github.com/kubernetes-sigs/jobset/pull/872/files#diff-4bc91a74bc47b8467cdcd25a7e3fe1651dbe44907f478377eca31b982f444f23R64-R79)
> 
> ### Controller Update:
> * [`pkg/controllers/jobset_controller.go`](https://github.com/kubernetes-sigs/jobset/pull/872/files#diff-68677f8530f00fbbb399ac36268f928ee584948dd09091644a8fad81a8658153R259): Updated the `updateJobSetStatus` function to call `RecordJobSetTerminalState` when updating the JobSet status. Added the `RecordJobSetTerminalState` function to check and record the terminal state of a JobSet. [[1]](https://github.com/kubernetes-sigs/jobset/pull/872/files#diff-68677f8530f00fbbb399ac36268f928ee584948dd09091644a8fad81a8658153R259) [[2]](https://github.com/kubernetes-sigs/jobset/pull/872/files#diff-68677f8530f00fbbb399ac36268f928ee584948dd09091644a8fad81a8658153R275-R293)
> 
> ### Tests:
> * [`pkg/metrics/metrics_test.go`](https://github.com/kubernetes-sigs/jobset/pull/872/files#diff-c3ade6c0a09b6258d971210f5a41c2ae5beb77873cee44a9ccb78055caf271ddR65-R84): Added a new test `TestRecordJobSetTerminalState` to verify that the `RecordJobSetTerminalState` function correctly sets the gauge metric for different terminal states.

This will make reviewers much more convenient! 😄

googs1025 commented:
> Code looks good to me.
> 
> I'll wait until docs are there.

done

kannon92 commented:
/approve

/cc @ahg-g @danielvegamyhre 

Either of you want to take a look?

kannon92 commented:
oh, I missed this.

Can you add a release note?

This is a feature and should have a release note.

kannon92 commented:
/hold

googs1025 commented:
> oh, I missed this.
> 
> Can you add a release note?
> 
> This is a feature and should have a release note.

done

googs1025 commented:
/unhold

kannon92 commented:
/hold
/approve cancel

thinking about this more, are we concerned with this overloading Prometheus with cardinals? Every JobSet would be populated in metrics which could be quite a large amount of data.

why would this be better in metrics than just having someone use kubectl to view the jobsets?

kannon92 commented on pkg/metrics/metrics.go:
```diff
@@ -39,6 +39,14 @@ var (
+			Subsystem: constants.JobSetSubsystemName,
+			Name:      "terminal_state",
+			Help:      `The current number of JobSets in terminal state (e.g. Completed, Failed)`,
+		}, []string{"jobset_name", "namespace", "terminal_state"},
```
We should follow the pattern for the other metrics. We combine namespace/name for label. Let’s follow that here.

googs1025 commented on site/content/en/docs/reference/metrics.md:
```diff
@@ -28,7 +28,8 @@ Use the following metrics to monitor the health of the jobset controller:
-| `jobset_completed_total`                    | Counter | The total number of completed JobSets.                                    | `jobset_name`: name of jobset  |
+| Metric Name                                | Type    | Description                                                                                     | Labels                                                                                             |
+|--------------------------------------------|---------|-------------------------------------------------------------------------------------------------|----------------------------------------------------------------------------------------------------|
+| `jobset_failed_total`                      | Counter | The total number of failed JobSets.                                                            | `jobset_name`: Identifier for the JobSet in the format `<namespace>/<name>`, where `<namespace>` is the namespace and `<name>` is the name of the JobSet. |
```
change other metrics describe

kannon92 commented:
So:

```
var (
	FailedTotal = prometheus.NewCounterVec(
		prometheus.CounterOpts{
			Subsystem: constants.JobSetSubsystemName,
			Name:      "failed_total",
			Help:      `The total number of failed JobSets`,
		}, []string{"jobset_name"},
	)

	CompletedTotal = prometheus.NewCounterVec(
		prometheus.CounterOpts{
			Subsystem: constants.JobSetSubsystemName,
			Name:      "completed_total",
			Help:      `The total number of completed JobSets`,
		}, []string{"jobset_name"},
	)

	JobSetTerminalState = prometheus.NewGaugeVec(
		prometheus.GaugeOpts{
			Subsystem: constants.JobSetSubsystemName,
			Name:      "terminal_state",
			Help:      `The current number of JobSets in terminal state (e.g. Completed, Failed)`,
		}, []string{"jobset_name", "terminal_state"},
	)
)
```

It seems that your metric duplicates FailedTotal and CompletedTotal.

I am not sure if we should have metrics that duplicate other metrics. 

WDYT @ahg-g @tenzen-y?

googs1025 commented:
Gaugemetric and the Counter metric are different purposes.  🤔  At least, this is how we define them internally.

Gauge represents an instantaneous value, indicating the current state.
It is used for right now monitoring of whether a JobSet is currently in a terminal state (e.g., Completed or Failed).

For example:
```yaml
jobset_terminal_state{jobset_name="default/example-jobset", terminal_state="Failed"}
```

Counter is used to track historical of JobSet, such as failure rates, which can be very useful. For example, sometimes our department's training tasks require calculating the failure rate over a week, which can use as an operational metric.

Gauge tracks the current number of JobSets in a terminal state. This metric is more relevant to sre , who can use it to troubleshoot issues or take some actions based on whether failures have occurred in real time.

These two metrics differ in meaning and purpose. If we can support both simultaneously, I believe it would be great! 😄

kannon92 commented:
I guess I am a bit confused on the fact that the counter metrics also have a jobset_name label.

```
jobset_completed{jobset_name="default/example-jobset"}
```

So would this be counting how many times that jobset completed?

googs1025 commented:
> I guess I am a bit confused on the fact that the counter metrics also have a jobset_name label.
> 
> ```
> jobset_completed{jobset_name="default/example-jobset"}
> ```
> 
> So would this be counting how many times that jobset completed?

thanks @kannon92 for quick reply.
Yes, I understand that this is a redundant field when I think about it. we can filter by label or not.

googs1025 commented:
But I understand that maybe we should distinguish <jobset>, <namespace>
as two labels?

googs1025 commented:
/close 

This seems like duplication of work
